lib/autoupdate/client_update.go

```diff
@@ -0,0 +1,353 @@
+	extractDir := filepath.Join(u.toolsDir, pkgName)
+	if runtime.GOOS != constants.DarwinOS {
+		if err := os.Mkdir(extractDir, 0o755); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	// Perform atomic replace so concurrent exec do not fail.
+	if err := packaging.ReplaceToolsBinaries(u.toolsDir, archivePath, extractDir, u.tools); err != nil {
+		return trace.Wrap(err)
+	}
+	// Cleanup the tools directory with previously downloaded and un-archived versions.
+	if err := packaging.RemoveWithSuffix(u.toolsDir, updatePackageSuffix, pkgName); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// Exec re-executes tool command with same arguments and environ variables.
+func (u *ClientUpdater) Exec() (int, error) {
+	path, err := toolName(u.toolsDir)
+	if err != nil {
+		return 0, trace.Wrap(err)
+	}
+
+	cmd := exec.Command(path, os.Args[1:]...)
+	// To prevent re-execution loop we have to disable update logic for re-execution.
+	cmd.Env = append(os.Environ(), teleportToolsVersionEnv+"=off")
+	cmd.Stdin = os.Stdin
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+
+	if err := cmd.Run(); err != nil {
+		return 0, trace.Wrap(err)
+	}
+
+	return cmd.ProcessState.ExitCode(), nil
+}
+
+func (u *ClientUpdater) downloadHash(ctx context.Context, url string) (string, error) {
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	resp, err := u.client.Do(req)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return "", trace.BadParameter("bad status when downloading archive hash: %v", resp.StatusCode)
+	}
+
+	var buf bytes.Buffer
+	_, err = io.CopyN(&buf, resp.Body, checksumHexLen)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	raw := buf.String()
+	if _, err = hex.DecodeString(raw); err != nil {
+		return "", trace.Wrap(err)
+	}
+	return raw, nil
+}
+
+func (u *ClientUpdater) downloadArchive(ctx context.Context, downloadDir string, url string, hash string) (string, error) {
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	resp, err := u.client.Do(req)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return "", trace.BadParameter("bad status when downloading archive: %v", resp.StatusCode)
+	}
+
+	if resp.ContentLength != -1 {
+		if err := checkFreeSpace(u.toolsDir, uint64(resp.ContentLength)); err != nil {
+			return "", trace.Wrap(err)
+		}
+	}
+
+	// Caller of this function will remove this file after the atomic swap has
+	// occurred.
+	f, err := os.CreateTemp(downloadDir, "tmp-")
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	h := sha256.New()
+	pw := &progressWriter{n: 0, limit: resp.ContentLength}
+	body := cancelableTeeReader(io.TeeReader(resp.Body, h), pw, syscall.SIGINT, syscall.SIGTERM)
```

Comment:
> context should be canceled by signal anyway if we want to stop downloading

Exactly, better to use the context abstraction, instead of hooking into the OS at this level